Add documentation page for typing-extensions (#166)
python/typing_extensions: doc/index.rst

.. data:: ClassVar

   See :py:data:`typing.ClassVar`. In ``typing`` since 3.5.3.

.. data:: Concatenate

   See :py:data:`typing.Concatenate`. In ``typing`` since 3.10.

.. data:: Final

   See :py:data:`typing.Final`. In ``typing`` since 3.8.

.. data:: Literal

   See :py:data:`typing.Literal`. In ``typing`` since 3.8.

   :py:data:`typing.Literal` does not flatten or deduplicate parameters on Python <3.9.1, and a
   caching bug was fixed in 3.10.1/3.9.8. The ``typing_extensions`` version
   flattens and deduplicates parameters on all Python versions, and the caching
   bug is also fixed on all versions.

   .. versionchanged:: 4.6.0

      Backported the bug fixes from :py:gh:`29334`, :py:gh:`23294`, and :py:gh:`23383`.

.. data:: LiteralString

   See :py:data:`typing.LiteralString`. In ``typing`` since 3.11.

   .. versionadded:: 4.1.0

.. class:: NamedTuple

   See :py:class:`typing.NamedTuple`.

   ``typing_extensions`` backports several changes
   to ``NamedTuple`` on Python 3.11 and lower: in 3.11,
   support for generic ``NamedTuple`` was added, and
   in 3.12, the ``__orig_bases__`` attribute was added.

   .. versionadded:: 4.3.0

>       Added to provide support for generic ``NamedTuple``.

Reviewer: ```suggestion
      Added to provide support for generic ``NamedTuple``s.
```

Author: ```suggestion
      Added to provide support for generic ``NamedTuple``\ s.
```

   .. versionchanged:: 4.6.0

      Support for the ``__orig_bases__`` attribute was added.

.. data:: Never

   See :py:data:`typing.Never`. In ``typing`` since 3.11.

   .. versionadded:: 4.1.0

.. class:: NewType(name, tp)

   See :py:class:`typing.NewType`. In ``typing`` since 3.5.2.

>    ``NewType`` was made picklable in 3.10 and an error message was

Reviewer: ```suggestion
   Instances of ``NewType`` were made picklable in 3.10 and an error message was
```

   .. versionchanged:: 4.6.0

      The improvements from Python 3.10 and 3.11 were backported.

.. data:: NoReturn

   See :py:data:`typing.NoReturn`. In ``typing`` since 3.5.4 and 3.6.2.

.. data:: NotRequired

   See :py:data:`typing.NotRequired`. In ``typing`` since 3.11.

   .. versionadded:: 4.0.0

.. class:: ParamSpec(name, *, default=...)

   See :py:class:`typing.ParamSpec`. In ``typing`` since 3.10.

   The ``typing_extensions`` version adds support for the
   ``default=`` argument from :pep:`696`.

   .. versionchanged:: 4.4.0

      Added support for the ``default=`` argument.

   .. versionchanged:: 4.6.0

      The implementation was changed for compatibility with Python 3.12.

.. class:: ParamSpecArgs

.. class:: ParamSpecKwargs

   See :py:class:`typing.ParamSpecArgs` and :py:class:`typing.ParamSpecKwargs`.
   In ``typing`` since 3.10.

.. class:: Protocol

   See :py:class:`typing.Protocol`. In ``typing`` since 3.8.

   Python 3.12 improves the performance of runtime-checkable protocols;
   ``typing_extensions`` backports this improvement.

   .. versionchanged:: 4.6.0

      Backported the ability to define ``__init__`` methods on Protocol classes.

>    .. versionchanged:: 4.6.0

Reviewer: Could possibly also mention https://github.com/python/typing_extensions/pull/140 and #132 in versionchanged notes for `Protocol` — they bring the implementation in line with Python 3.12, and they're strictly-speaking bugfixes, but they're also minor compatibility breaks with previous versions

Author: Done, thanks for all the comments!

>       Backported changes to runtime-checkable protocols from Python 3.12.

Author: ```suggestion
      Backported changes to runtime-checkable protocols from Python 3.12,
      including :py:gh:`103034` and :py:gh:`26067`.
```

   See :py:class:`typing.Type`. In ``typing`` since 3.5.2.

.. data:: TypeAlias

   See :py:data:`typing.TypeAlias`. In ``typing`` since 3.10.

.. class:: TypeAliasType(name, value, *, type_params=())

   See :py:class:`typing.TypeAliasType`. In ``typing`` since 3.12.

   .. versionadded:: 4.6.0

.. data:: TypeGuard

   See :py:data:`typing.TypeGuard`. In ``typing`` since 3.10.

.. class:: TypedDict

   See :py:class:`typing.TypedDict`. In ``typing`` since 3.8.

   ``typing_extensions`` backports various bug fixes and improvements
   to ``TypedDict`` on Python 3.11 and lower.
   :py:class:`TypedDict` does not store runtime information
   about which (if any) keys are non-required in Python 3.8, and does not
   honor the ``total`` keyword with old-style ``TypedDict()`` in Python
   3.9.0 and 3.9.1. :py:class:`typing.TypedDict` also does not support multiple inheritance
   with :py:class:`typing.Generic` on Python <3.11, and :py:class:`typing.TypedDict` classes do not
   consistently have the ``__orig_bases__`` attribute on Python <3.12. The
   ``typing_extensions`` backport provides all of these features and bugfixes on
   all Python versions.

   .. versionadded:: 4.3.0

>       Added to provide support for generic ``TypedDict``.

Reviewer: I don't think that's true — it existed in typing_extensions before we modified the implementation to provide support for generic TypedDicts

Author: ```suggestion
      Added support for generic ``TypedDict``\ s.
```
Oops yes, confused myself with NamedTuple

Reviewer: It needs to be "versionchanged" as well as "versionadded", it existed in `typing_extensions` before 4.3.0. I don't know when it was added.

   .. versionchanged:: 4.6.0

      A :py:exc:`DeprecationWarning` is now emitted when a call-based
      ``TypedDict`` is constructed using keyword arguments.

   .. versionchanged:: 4.6.0

      Support for the ``__orig_bases__`` attribute was added.

.. class:: TypeVar(name, *constraints, bound=None, covariant=False,
                   contravariant=False, infer_variance=False, default=...)

   See :py:class:`typing.TypeVar`.

   The ``typing_extensions`` version adds support for the
   ``default=`` argument from :pep:`696`, as well as the
   ``infer_variance=`` argument from :pep:`695` (also available
   in Python 3.12).

   .. versionadded:: 4.4.0

      Added in order to support the new ``default=`` and
      ``infer_variance=`` arguments.

   .. versionchanged:: 4.6.0

      The implementation was changed for compatibility with Python 3.12.

.. class:: TypeVarTuple(name, *, default=...)

>    See :py:class:`typing.TypeVarTuple`. In ``typing`` since 3.10.

Reviewer: ```suggestion
   See :py:class:`typing.TypeVarTuple`. In ``typing`` since 3.11.
```

   The ``typing_extensions`` version adds support for the
   ``default=`` argument from :pep:`696`.

   .. versionadded:: 4.1.0

   .. versionchanged:: 4.4.0

      Added support for the ``default=`` argument.

   .. versionchanged:: 4.6.0

      The implementation was changed for compatibility with Python 3.12.

.. data:: Unpack

>    See :py:data:`typing.Unpack`. In ``typing`` since 3.10.

Reviewer: ```suggestion
   See :py:data:`typing.Unpack`. In ``typing`` since 3.11.
```

.. class:: Deque

   See :py:class:`typing.Deque`. In ``typing`` since 3.5.4 and 3.6.1.

.. class:: OrderedDict

   See :py:class:`typing.OrderedDict`. In ``typing`` since 3.7.2.

Abstract Base Classes
~~~~~~~~~~~~~~~~~~~~~

.. class:: AsyncContextManager

   See :py:class:`typing.AsyncContextManager`. In ``typing`` since 3.5.4 and 3.6.2.

.. class:: AsyncGenerator

   See :py:class:`typing.AsyncGenerator`. In ``typing`` since 3.6.1.

.. class:: AsyncIterable

   See :py:class:`typing.AsyncIterable`. In ``typing`` since 3.5.2.

.. class:: AsyncIterator

   See :py:class:`typing.AsyncIterator`. In ``typing`` since 3.5.2.

.. class:: Awaitable

   See :py:class:`typing.Awaitable`. In ``typing`` since 3.5.2.

.. class:: Buffer

   See :py:class:`collections.abc.Buffer`. Added to the standard library
   in Python 3.12.

   .. versionadded:: 4.6.0

.. class:: ContextManager

   See :py:class:`typing.ContextManager`. In ``typing`` since 3.5.4.

.. class:: Coroutine

   See :py:class:`typing.Coroutine`. In ``typing`` since 3.5.3.

Protocols
~~~~~~~~~

.. class:: SupportsAbs

   See :py:class:`typing.SupportsAbs`.

   ``typing_extensions`` backports a more performant version of this

>    protocol in Python 3.11 and lower.

Reviewer: I think I'd go for "on" here (and in the other Protocol descriptions below)?

```suggestion
   protocol on Python 3.11 and lower.
```

   See :py:class:`typing.SupportsInt`.

   ``typing_extensions`` backports a more performant version of this
   protocol in Python 3.11 and lower.

   .. versionadded:: 4.6.0

.. class:: SupportsRound

   See :py:class:`typing.SupportsRound`.

   ``typing_extensions`` backports a more performant version of this
   protocol in Python 3.11 and lower.

   .. versionadded:: 4.6.0

Decorators
~~~~~~~~~~

.. decorator:: dataclass_transform(*, eq_default=False, order_default=False,
                                   kw_only_default=False, frozen_default=False,
                                   field_specifiers=(), **kwargs)

   See :py:func:`typing.dataclass_transform`. In ``typing`` since 3.11.

   Python 3.12 adds the ``frozen_default`` parameter; ``typing_extensions``
   backports this parameter.

   .. versionadded:: 4.1.0

   .. versionchanged:: 4.2.0

      The ``field_descriptors`` parameter was renamed to ``field_specifiers``.
      For compatibility, the decorator now accepts arbitrary keyword arguments.

   .. versionchanged:: 4.5.0

      The ``frozen_default`` parameter was added.

.. decorator:: deprecated(msg, *, category=DeprecationWarning, stacklevel=1)

>    See :pep:`702`.

Reviewer: ```suggestion
   See :pep:`702`. Experimental; not yet part of the standard library.
```

   .. versionadded:: 4.5.0

.. decorator:: final

   See :py:func:`typing.final`. In ``typing`` since 3.8.

   Since Python 3.11, this decorator supports runtime introspection

>    through the ``__final__`` attribute; ``typing_extensions.final``

Reviewer: ```suggestion
   by setting the ``__final__`` attribute wherever possible; ``typing_extensions.final``
```

   backports this feature.

   .. versionchanged:: 4.1.0

>       The ``__final__`` attribute is now set on decorated objects.

Reviewer: Maybe something like:

> The decorator now attempts to set the `__final__` attribute on decorated objects. AttributeErrors and TypeErrors are caught when trying to set this attribute.

Author: ```suggestion
      The decorator now attempts to set the ``__final__`` attribute on decorated objects. 
```
Left out the second sentence for brevity; details of behavior can be in the CPython docs.

.. decorator:: overload

   See :py:func:`typing.overload`.

   Since Python 3.11, this decorator supports runtime introspection
   through :func:`get_overloads`; ``typing_extensions.overload``
   backports this feature.

   .. versionchanged:: 4.2.0

      Added to provide support for introspection through :func:`get_overloads` was added.

.. decorator:: override

   See :py:func:`typing.override`. In ``typing`` since 3.12.

   .. versionadded:: 4.4.0

   .. versionchanged:: 4.5.0

>       The decorator now sets the ``__override__`` attribute on the decorated

Reviewer: See my comment above about `@final`

Author: ```suggestion
      The decorator now attempts to set the ``__override__`` attribute on the decorated
```

   .. versionadded:: 4.1.0

.. function:: assert_type(val, typ)

   See :py:func:`typing.assert_type`. In ``typing`` since 3.11.

   .. versionadded:: 4.2.0

.. function:: clear_overloads()

   See :py:func:`typing.clear_overloads`. In ``typing`` since 3.11.

   .. versionadded:: 4.2.0

.. function:: get_args(tp)

   See :py:func:`typing.get_args`. In ``typing`` since 3.8.

   This function was changed in 3.9 and 3.10 to deal with :data:`Annotated`
   and :class:`ParamSpec` correctly; ``typing_extensions`` backports these
   fixes.

.. function:: get_origin(tp)

   See :py:func:`typing.get_origin`. In ``typing`` since 3.8.

   This function was changed in 3.9 and 3.10 to deal with :data:`Annotated`
   and :class:`ParamSpec` correctly; ``typing_extensions`` backports these
   fixes.

.. function:: get_original_bases(cls)

   See :py:func:`types.get_original_bases`. Added to the standard library
   in Python 3.12.

   This function should always produce correct results when called on classes
   constructed using features from ``typing_extensions``. However, it may
   produce incorrect results when called on some :py:class:`NamedTuple` or

>    :py:class:`TypedDict` on Python <=3.11.

Reviewer: ```suggestion
   :py:class:`TypedDict` classes on Python <=3.11.
```

   .. versionadded:: 4.6.0

.. function:: get_overloads(func)

   See :py:func:`typing.get_overloads`. In ``typing`` since 3.11.

   Before Python 3.11, this works only with overloads created through
   :func:`overload`, not with :py:func:`typing.overload`.

   .. versionadded:: 4.2.0

.. function:: get_type_hints(obj, globalns=None, localns=None, include_extras=False)

   See :py:func:`typing.get_type_hints`.

   In Python 3.11, this function was changed to support the new
   :py:data:`typing.Required`. ``typing_extensions`` backports these
   fixes.

>    .. versionchanged:: 4.1.0

Reviewer: Presumably we must also have backported the `include_extras` parameter at some point after PEP 593 was implemented in Python 3.9

Author: I think that was before we started a changelog. I haven't gone through the git archeology to figure out version changes from that time.

Reviewer: Yeah, I tried digging through the git archaeology, and quickly gave up :)

      Interaction with :data:`Required` and :data:`NotRequired`.

.. function:: is_typeddict(tp)

   See :py:func:`typing.is_typeddict`. In ``typing`` since 3.10.

   On versions where :class:`TypedDict` is not the same as
   :py:class:`typing.TypedDict`, this function recognizes

>    ``TypedDict`` classes created through either.

Reviewer: ```suggestion
   ``TypedDict`` classes created through either mechanism.
```

Welcome to typing_extensions's documentation!
=============================================

> ``typing_extensions`` provides common building blocks for type annotations
> that are understood by static type checkers.
> 
> The ``typing_extensions`` module serves two related purposes:

Reviewer: It would be nice to mention (and link to) the `typing` module in the first paragraph, I think. Maybe something like:

```suggestion
``typing_extensions`` complements the standard-library :py:mod:`typing` module,
providing runtime support for type hints as specified by PEP 484 and subsequent
PEPs. The module serves two related purposes:
```

Author: It would be nice to mention (and link to) the `typing` module in the first paragraph, I think. Maybe something like:
Also add a link:
```suggestion
``typing_extensions`` complements the standard-library :py:mod:`typing` module,
providing runtime support for type hints as specified by :pep:`484` and subsequent
PEPs. The module serves two related purposes:
```

- Enable use of new type system features on older Python versions. For example,
  :py:data:`typing.TypeGuard` is new in Python 3.10, but ``typing_extensions`` allows
  users on previous Python versions to use it too.

> - Enable experimentation with new type system PEPs before they are accepted and
>   added to the :py:mod:`typing` module.

Reviewer: "Type system PEPs" sounds slightly strange to me -- an alternative solution would be "type-system PEPs", but that also feels slightly awkward

```suggestion
- Enable experimentation with new PEPs relating to the type system before they
  are accepted and added to the :py:mod:`typing` module.
```

Author: How about:

```suggestion
- Enable experimentation with type system features proposed in new PEPs before they are accepted and
  added to the :py:mod:`typing` module.
```

Reviewer: Yes, that works well!

New features may be added to ``typing_extensions`` as soon as they are specified
in a PEP that has been added to the `python/peps <https://github.com/python/peps>`_

> repository. If the PEP is accepted, the feature will then be added to :py:mod:`typing`
> for the next CPython release. No typing PEP that affected ``typing_extensions``
> has been rejected so far, so we haven't yet figured out how to deal with that possibility.

Reviewer: ```suggestion
repository. If the PEP is accepted, the feature will then be added to the
:py:mod:`typing` module for the next CPython release. No typing PEP that
affected ``typing_extensions`` has been rejected so far, so we haven't yet
figured out how to deal with that possibility.
```

.. class:: Type

   See :py:class:`typing.Type`. In ``typing`` since 3.5.2.

.. data:: TypeAlias

   See :py:data:`typing.TypeAlias`. In ``typing`` since 3.10.

.. class:: TypeAliasType(name, value, *, type_params=())

   See :py:class:`typing.TypeAliasType`. In ``typing`` since 3.12.

   .. versionadded:: 4.6.0

.. data:: TypeGuard

   See :py:data:`typing.TypeGuard`. In ``typing`` since 3.10.

.. class:: TypedDict

   See :py:class:`typing.TypedDict`. In ``typing`` since 3.8.

   ``typing_extensions`` backports various bug fixes and improvements
   to ``TypedDict`` on Python 3.11 and lower.
   :py:class:`TypedDict` does not store runtime information
   about which (if any) keys are non-required in Python 3.8, and does not
   honor the ``total`` keyword with old-style ``TypedDict()`` in Python
   3.9.0 and 3.9.1. :py:class:`typing.TypedDict` also does not support multiple inheritance
   with :py:class:`typing.Generic` on Python <3.11, and :py:class:`typing.TypedDict` classes do not
   consistently have the ``__orig_bases__`` attribute on Python <3.12. The
   ``typing_extensions`` backport provides all of these features and bugfixes on
   all Python versions.

>    .. versionadded:: 4.3.0

Reviewer: Don't know when it was added, but it wasn't in 4.3.0 :)

```suggestion
   .. versionchanged:: 4.3.0
```

Author: Thanks, just pushed that

   Python 3.12 adds the ``frozen_default`` parameter; ``typing_extensions``
   backports this parameter.

   .. versionadded:: 4.1.0

   .. versionchanged:: 4.2.0

      The ``field_descriptors`` parameter was renamed to ``field_specifiers``.
      For compatibility, the decorator now accepts arbitrary keyword arguments.

   .. versionchanged:: 4.5.0

      The ``frozen_default`` parameter was added.

.. decorator:: deprecated(msg, *, category=DeprecationWarning, stacklevel=1)

   See :pep:`702`. Experimental; not yet part of the standard library.

   .. versionadded:: 4.5.0

.. decorator:: final

   See :py:func:`typing.final`. In ``typing`` since 3.8.

   Since Python 3.11, this decorator supports runtime introspection
   by setting the ``__final__`` attribute wherever possible; ``typing_extensions.final``
   backports this feature.

   .. versionchanged:: 4.1.0

      The decorator now attempts to set the ``__final__`` attribute on decorated objects.

.. decorator:: overload

   See :py:func:`typing.overload`.

   Since Python 3.11, this decorator supports runtime introspection
   through :func:`get_overloads`; ``typing_extensions.overload``
   backports this feature.

>    .. versionchanged:: 4.2.0

Author: ```suggestion
   .. versionadded:: 4.2.0
```

>       Added to provide support for introspection through :func:`get_overloads` was added.

Reviewer: ```suggestion
      Introspection support via :func:`get_overloads` was added.
```

Author: There was no `typing_extensions.overload` before this, so the note is here to explain why we added. I should actually change the line above to "versionadded", not "versionchanged".

Reviewer: Are you sure? I can see it as far back as 3.10.1: https://github.com/python/typing_extensions/blob/0c70227e221866b74b5cf6b3e251bc67864781ef/src_py3/typing_extensions.py#L693-L719

Author: Oh you're right, I misremembered. I wonder why; according to https://docs.python.org/3.10/library/typing.html#typing.overload it's always been in `typing`.

Reviewer: The past is a foreign country: they do things differently there

   fixes.

.. function:: get_origin(tp)

   See :py:func:`typing.get_origin`. In ``typing`` since 3.8.

   This function was changed in 3.9 and 3.10 to deal with :data:`Annotated`
   and :class:`ParamSpec` correctly; ``typing_extensions`` backports these
   fixes.

.. function:: get_original_bases(cls)

   See :py:func:`types.get_original_bases`. Added to the standard library
   in Python 3.12.

   This function should always produce correct results when called on classes
   constructed using features from ``typing_extensions``. However, it may
   produce incorrect results when called on some :py:class:`NamedTuple` or
   :py:class:`TypedDict` classes on Python <=3.11.

   .. versionadded:: 4.6.0

.. function:: get_overloads(func)

   See :py:func:`typing.get_overloads`. In ``typing`` since 3.11.

   Before Python 3.11, this works only with overloads created through
   :func:`overload`, not with :py:func:`typing.overload`.

   .. versionadded:: 4.2.0

.. function:: get_type_hints(obj, globalns=None, localns=None, include_extras=False)

   See :py:func:`typing.get_type_hints`.

   In Python 3.11, this function was changed to support the new

>    :py:data:`typing.Required`. ``typing_extensions`` backports these
>    fixes.

Reviewer: ```suggestion
   :py:data:`typing.Required` and :py:data:`typing.NotRequired`.
   ``typing_extensions`` backports these fixes.
```

   .. versionchanged:: 4.1.0

      Interaction with :data:`Required` and :data:`NotRequired`.

.. function:: is_typeddict(tp)

   See :py:func:`typing.is_typeddict`. In ``typing`` since 3.10.

   On versions where :class:`TypedDict` is not the same as
   :py:class:`typing.TypedDict`, this function recognizes
   ``TypedDict`` classes created through either mechanism.

   .. versionadded:: 4.1.0

.. function:: reveal_type(obj)

   See :py:func:`typing.reveal_type`. In ``typing`` since 3.11.

   .. versionadded:: 4.1.0

Other
~~~~~

.. class:: Text

   See :py:class:`typing.Text`. In ``typing`` since 3.5.2.

.. data:: TYPE_CHECKING

   See :py:data:`typing.TYPE_CHECKING`. In ``typing`` since 3.5.2.

Limitations
-----------

> There are a few types whose interface was modified between different

Reviewer: Or, if we keep "whose", then "interface" needs to be plural ("interfaces were" rather than "interface was")

```suggestion
There are a few types where the interface was modified between different
```

versions of typing. For example, :py:class:`typing.Sequence` was modified to
subclass :py:class:`typing.Reversible` as of Python 3.5.3.

These changes are *not* backported to prevent subtle compatibility
issues when mixing the differing implementations of modified classes.

> Certain types have incorrect runtime behavior due to limitations of older
> versions of the typing module:

Reviewer: Make it slightly clearer that this is a distinct point to the previous paragraph:

```suggestion
As well as this, certain types have incorrect runtime behavior due to
limitations of older versions of the typing module:
```

Author: I actually think we should get rid of this catch-all section and document these limitations only under specific objects. I'll move a version of this text to the ParamSpec docs instead, and get rid of the sentence about Sequence/Reversible which I think is only relevant to 3.5-era changes.

Reviewer: Yes, that makes sense. Sounds like a plan.

Author: I couldn't figure out what's supposedly wrong with Concatenate, so I'm going to add a note to ParamSpec only.

```
Python 3.8.16 (default, Dec 26 2022, 20:36:56) 
[Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing_extensions import Concatenate, get_args, get_origin, ParamSpec
>>> P = ParamSpec("P")
>>> c = Concatenate[int, P]
>>> get_origin(c)
typing_extensions.Concatenate
>>> get_args(c)
(<class 'int'>, ~P)
```

Reviewer: > I couldn't figure out what's supposedly wrong with Concatenate, so I'm going to add a note to ParamSpec only.

Makes sense, though #110 might be worth a mention on the Concatenate entry. I took a brief look at it a while back, and it looks... challenging to backport.

Reviewer: #48 as well...

Author: Good point, added those. I also noticed that the :gh: links to CPython weren't working, so I added a custom Sphinx extension to make them work (possibly overkill?).

Reviewer: > possibly overkill?

Definitely a Choice, but I'll support you in it
